File: 2022/go/day14/day14.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day14.txt")

	fmt.Printf("part1: %v\n", pourSand(input, false))
	fmt.Printf("part2: %v\n", pourSand(input, true))
}

func pourSand(input []byte, useFoor bool) (count int) {
	cave := readRocks(string(input))
	floor := 0
	if useFoor {
		floor = bottom(cave)
	}
	var start = xy{500, 0}
	count = -1
	for count != countSand(cave) {
		count = countSand(cave)
		dropSand(cave, floor, start)
	}
	return count
}

func countSand(cave map[xy]byte) (count int) {
	for _, i := range cave {
		if i == 'o' {
			count++
		}
	}
	return count
}

func dropSand(cave map[xy]byte, floor int, sand xy) {
	if sand.y != floor-1 && sand.y < 1000 { // Try if sand can be moved down
		for _, v := range moves {
			try := xy{sand.x + v.x, sand.y + v.y}
			if _, found := cave[try]; !found {
				dropSand(cave, floor, try) // can be moved, try next
				return
			}
		}
	}

	if _, support := cave[xy{sand.x, sand.y + 1}]; support || sand.y+1 == floor {
		cave[sand] = 'o' // support below, place sand
	}
}

func bottom(cave map[xy]byte) (res int) {
	for p := range cave {
		if p.y > res {
			res = p.y
		}
	}
	return res + 2
}

var moves = []xy{{0, 1}, {-1, 1}, {1, 1}}

func readRocks(input string) map[xy]byte {
	lines := strings.Split(input, "\n")
	rocks := map[xy]byte{}
	for _, line := range lines {
		points := strings.Split(line, " -> ")
		for i := 1; i < len(points); i++ {
			var p1, p2 xy
			fmt.Sscanf(points[i-1], "%d,%d", &p1.x, &p1.y)
			fmt.Sscanf(points[i], "%d,%d", &p2.x, &p2.y)
			draw(rocks, p1, p2)
		}
	}
	return rocks
}

func draw(rocks map[xy]byte, p1, p2 xy) {
	rocks[p1] = '#'
	if p1 == p2 {
		return
	}

	if p1.x == p2.x && p1.y < p2.y {
		draw(rocks, xy{p1.x, p1.y + 1}, p2)
	} else if p1.x == p2.x && p1.y > p2.y {
		draw(rocks, xy{p1.x, p1.y - 1}, p2)
	} else if p1.y == p2.y && p1.x < p2.x {
		draw(rocks, xy{p1.x + 1, p1.y}, p2)
	} else if p1.y == p2.y && p1.x > p2.x {
		draw(rocks, xy{p1.x - 1, p1.y}, p2)
	}
}

type xy struct{ x, y int }

File: 2022/java/Day14.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class Day14 {

    public static void main(String ... args) throws Exception {
        System.out.printf("part1: %d\n", pourSand(false));
        System.out.printf("part2: %d\n", pourSand(true));
    }

    static long pourSand(boolean haveFloor) throws Exception {
        var cave = readRocks(Files.readString(Path.of("../input/day14.txt")));
        int floor = haveFloor ? cave.keySet().stream().mapToInt(r -> r.y).max().getAsInt() + 2 : 0;
        long count = -1;
        while(count != countSand(cave)) {
            count = countSand(cave);
            dropSand(cave, floor);
        }
        return count;
    }

    static void dropSand(Map<XY,Entry> cave, int floor) {
        XY sand = new XY(500, 0), prev;
        do {
            prev = sand;
            XY old = sand;
            sand = moves.stream().map(m -> new XY(old.x + m.x, old.y + m.y))
                .filter(n -> !cave.containsKey(n)).findFirst().orElse(sand); // find next place for sand
        } while (prev != sand && sand.y + 1 != floor && sand.y < 999);

        if (cave.containsKey(new XY(sand.x, sand.y+1)) || sand.y + 1 == floor) {
            cave.put(sand, Entry.SAND); // We have support below, place sand
        }
    }

    static long countSand(Map<XY,Entry> cave) {
        return cave.values().stream().filter(e -> e == Entry.SAND).count();
    }

    static Map<XY,Entry> readRocks(String input) {
        return Arrays.stream(input.split("\n")).flatMap(l -> toRocks(l).stream())
            .collect(Collectors.toMap(r -> r, r -> Entry.ROCK, (a,b) -> a, () -> new HashMap<XY,Entry>()));
    }

    static List<XY> toRocks(String line) {
        String[] parts = line.split(" -> ");
        var res = new ArrayList<XY>();
        for (int i = 1; i < parts.length; i++) {
            res.addAll(draw(XY.parse(parts[i-1]), XY.parse(parts[i])).toList());
        }
        return res;
    }

    static Stream<XY> draw(XY p1, XY p2) {
        if (p1.equals(p2)) { return Stream.of(p1); }
        return Stream.concat(Stream.of(p1), draw(p1.moveTowards(p2), p2));
    }

    enum Entry{ROCK, SAND};

    record XY(int x, int y) {
        static XY parse(String input) {
            var parts = input.split(",");
            return new XY(Integer.parseInt(parts[0]), Integer.parseInt(parts[1]));
        }
        XY moveTowards(XY o) {
            if (x == o.x && y < o.y) return new XY(x, y+1);
            if (x == o.x && y > o.y) return new XY(x, y-1);
            if (y == o.y && x < o.x) return new XY(x+1, y);
            if (y == o.y && x > o.x) return new XY(x-1, y);
            throw new IllegalStateException();
        }
    }

    static List<XY> moves = List.of(new XY(0, 1), new XY(-1, 1), new XY(1, 1));
 }